File: Contests/14/H.py
n, m = map(int, input().split())
b = [0] * (n+1)

for i in range(m):
    u, v = map(int, input().split())
    b[u] = b[u] | 1 << v
    b[v] = b[v] | 1 << u

t = False
for i in range(1, n + 1):
    for j in range(i + 1, n + 1):
        if bin(b[i] & b[j]).count('1') > 2:
            t = True
            break
print("YES" if t else "NO")

File: Contests/14/H.cpp
#include <iostream>
#include <vector>
#include <bitset>
using namespace std;

int main() {
    int n, m;
    cin >> n >> m;
    
    vector<int> b(3000, 0);
    
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        b[u] |= 1 << v;
        b[v] |= 1 << u;
    }
    
    bool t = false;
    for (int i = 1; i <= n; i++) {
        for (int j = i + 1; j <= n; j++) {
            if (bitset<32>(b[i] & b[j]).count() > 2) {
                t = true;
                break;
            }
        }
        if (t) {
            break;
        }
    }
    
    cout << (t ? "YES" : "NO") << endl;
    
    return 0;
}
